daemon/daemon.go

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package daemon
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/megaease/easeprobe/global"
+)
+
+// Config is the daemon config
+type Config struct {
+	PIDFile string
+	pidFd   *os.File
+}
+
+// NewPIDFile create a new pid file
+func NewPIDFile(pidfile string) (*Config, error) {
+	if pidfile == "" {
+		return nil, fmt.Errorf("pid file is empty")
+	}
+
+	c := &Config{
+		PIDFile: pidfile,
+	}
+	_, err := c.CheckPIDFile()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(c.PIDFile), 0755); err != nil {
+		return nil, err
+	}
+
+	pidstr := fmt.Sprintf("%d", os.Getpid())
+	if err := os.WriteFile(c.PIDFile, []byte(pidstr), 0600); err != nil {
+		return nil, err
+	}
+
+	c.pidFd, _ = os.OpenFile(c.PIDFile, os.O_APPEND|os.O_EXCL, 0600)
+	return c, nil
+}
+
+// CheckPIDFile check if the pid file exists
+// if the PID file exists, return the PID of the process
+// if the PID file does not exist, return -1
+func (c *Config) CheckPIDFile() (int, error) {
```

Comment:
Suggestion to also check that the pid file is indeed a file and not symlink or whatnot?

Reply:
made this change.